### PR TITLE
Add support for external Service registration

### PR DESCRIPTION
**Pull request by B3rn475**
Updated `service_undocumented.md` with the new Streams, Methods, and Events.

It was required to change the generator code to support Streams with `_` in the name and Type and Enum overrides.

### CODE REVIEW

**Comment by B3rn475** [comment]
@devoncarew

File: dart/tool/service_undocumented.md [hunk 1]
```diff
@@ -158,3 +158,53 @@ class HeapSpace extends Response {
...
+  String service [optional];
+
+  // The RPC method that should be used to invoke the service.
+  //
+  // This is provided for the event kinds:
+  //   ServiceRegistered
+  //   ServiceUnregistered
+  String method [optional];
+
+  // The alias of the registred service.
```
Comment (devoncarew) [inline dart/tool/service_undocumented.md]: sp: registered

File: dart/tool/dart/generate_dart.dart [hunk 2]
```diff
@@ -451,6 +456,30 @@ Stream<Event> onEvent(String streamName) => _getEventController(streamName).stre
     types.where((t) => !t.skip).forEach((t) => t.generate(gen));
   }
 
+  void _mergeTypes() {
+    final map = {};
```
Comment (devoncarew) [inline dart/tool/dart/generate_dart.dart]: Can you add types here on the LHS to match the surrounding code? (`<String, Type>` ?)

File: dart/tool/dart/generate_dart.dart [hunk 3]
```diff
@@ -451,6 +456,30 @@ Stream<Event> onEvent(String streamName) => _getEventController(streamName).stre
...
+        map[t.name] = new Type.merge(map[t.name], t);
+      } else {
+        map[t.name] = t;
+      }
+    }
+    types = map.values.toList();
+  }
+
+  void _mergeEnums() {
+    final map = {};
```
Comment (devoncarew) [inline dart/tool/dart/generate_dart.dart]: here as well (types)

File: dart/tool/java/generate_java.dart [hunk 4]
```diff
@@ -235,6 +239,30 @@ class Api extends Member with ApiParseUtil {
     }
   }
 
+  void _mergeTypes() {
+    final map = {};
```
Comment (devoncarew) [inline dart/tool/java/generate_java.dart]: types

File: dart/tool/java/generate_java.dart [hunk 5]
```diff
@@ -235,6 +239,30 @@ class Api extends Member with ApiParseUtil {
...
+        map[t.name] = new Type.merge(map[t.name], t);
+      } else {
+        map[t.name] = t;
+      }
+    }
+    types = map.values.toList();
+  }
+
+  void _mergeEnums() {
+    final map = {};
```
Comment (devoncarew) [inline dart/tool/java/generate_java.dart]: types

File: dart/tool/java/generate_java.dart [hunk 6]
```diff
@@ -272,13 +300,19 @@ class Api extends Member with ApiParseUtil {
       } else if (isPara(node)) {
         var children = (node as Element).children;
         if (children.isNotEmpty && children.first is Text) {
-          var text = (children.first as Text).text;
+          var text = children.expand<String>((var child) {
```
Comment (devoncarew) [inline dart/tool/java/generate_java.dart]: for this closure, you could either elide the `var` (for conciseness) or add the type of `child`

File: dart/tool/java/generate_java.dart [hunk 7]
```diff
@@ -272,13 +300,19 @@ class Api extends Member with ApiParseUtil {
...
+            if (child is Text)
```
Comment (devoncarew) [inline dart/tool/java/generate_java.dart]: add braces, or return on the same line as the if keyword

File: dart/tool/service_undocumented.md [hunk 8]
```diff
@@ -158,3 +158,53 @@ class HeapSpace extends Response {
 <!-- _getCpuProfileTimeline -->
 
 <!-- _getAllocationSamples -->
+
+streamId | event types provided
```
Comment (devoncarew) [inline dart/tool/service_undocumented.md]: Ah, I see why you have to merge above - portions of `EventKind` and `Event` are defined in two separate files. We could merge your change as is, but hopefully the split (some things defined in `service_undocumented.md`) is temporary. I'm wondering if there's not a simpler way to do this; in the parsing code, get `EventKind` and stick two new fields on it; get `Event` and stick 3 new fields on that.

Thoughts?

Comment (devoncarew) [inline dart/tool/service_undocumented.md]: I'm hoping that the plan is to promote `_registerService` and related to public API once things have been proved out.

Comment (B3rn475) [inline dart/tool/service_undocumented.md]: I think that the change will make things easier in the future even if we make the API public.

Right now the __Dart__ version supports extra stream and events that the __Java__ version does not, due to the fact that the parse and code generator does not support that and they are manually added before code generation.

File: dart/tool/service_undocumented.md (hunk 1 above)
Comment (B3rn475) [inline dart/tool/service_undocumented.md]: done

File: dart/tool/dart/generate_dart.dart (hunk 2 above)
Comment (B3rn475) [inline dart/tool/dart/generate_dart.dart]: done

File: dart/tool/dart/generate_dart.dart (hunk 3 above)
Comment (B3rn475) [inline dart/tool/dart/generate_dart.dart]: done

File: dart/tool/java/generate_java.dart (hunk 4 above)
Comment (B3rn475) [inline dart/tool/java/generate_java.dart]: done

File: dart/tool/java/generate_java.dart (hunk 5 above)
Comment (B3rn475) [inline dart/tool/java/generate_java.dart]: done

File: dart/tool/java/generate_java.dart (hunk 6 above)
Comment (B3rn475) [inline dart/tool/java/generate_java.dart]: done

File: dart/lib/vm_service_lib.dart [hunk 9]
```diff
@@ -731,36 +751,85 @@ class VmService {
...
+      }
+    } else if (json.containsKey('id') &&
+        (json.containsKey('result') || json.containsKey('error'))) {
+      _processResponse(json);
+    } else {
+      _log.severe('unknown message type: ${message}');
+    }
+  }
+
+  void _processResponse(json) {
```
Comment (devoncarew) [inline dart/lib/vm_service_lib.dart]: nit: add a type to `json` here (`dynamic`)

File: dart/lib/vm_service_lib.dart [hunk 10]
```diff
@@ -731,36 +751,85 @@ class VmService {
...
+      String type = result['type'];
+      if (_typeFactories[type] == null) {
+        completer.complete(Response.parse(result));
+      } else {
+        completer.complete(_createObject(result));
+      }
+    }
+  }
+
+  Future _processRequest(json) async {
```
Comment (devoncarew) [inline dart/lib/vm_service_lib.dart]: same re a type annotation

File: dart/lib/vm_service_lib.dart [hunk 11]
```diff
@@ -731,36 +751,85 @@ class VmService {
...
+    final Map m = await _routeRequest(json['method'], json['params']);
+    m['id'] = json.id;
+    m['jsonrpc'] = '2.0';
+    String message = JSON.encode(m);
+    _onSend.add(message);
+    _writeMessage(message);
+  }
+
+  Future _processNotification(json) async {
```
Comment (devoncarew) [inline dart/lib/vm_service_lib.dart]: ditto

File: dart/tool/dart/generate_dart.dart [hunk 12]
```diff
@@ -299,6 +358,8 @@ class Api extends Member with ApiParseUtil {
     } else {
       throw 'unexpected entity: ${name}, ${definition}';
     }
+    _mergeTypes();
```
Comment (devoncarew) [inline dart/tool/dart/generate_dart.dart]: Could you add a line doc here describing that we're merging the field for types and enums from the service.md and service_undocumented.md files?

File: dart/tool/service_undocumented.md (hunk 8 above)
Comment (devoncarew) [inline dart/tool/service_undocumented.md]: ok, sgtm

File: dart/tool/dart/generate_dart.dart (hunk 12 above)
Comment (B3rn475) [inline dart/tool/dart/generate_dart.dart]: done

File: dart/lib/vm_service_lib.dart (hunk 11 above)
Comment (B3rn475) [inline dart/lib/vm_service_lib.dart]: done

File: dart/lib/vm_service_lib.dart (hunk 10 above)
Comment (B3rn475) [inline dart/lib/vm_service_lib.dart]: done

File: dart/lib/vm_service_lib.dart (hunk 9 above)
Comment (B3rn475) [inline dart/lib/vm_service_lib.dart]: done

**Comment by devoncarew** [comment]
Thanks! landing

**Pull request merged**